dace/transformation/helpers.py

```diff
@@ -17,6 +18,307 @@
+
+        # Find NestedSDFG's connectors
+        read_set = {n for n in read_set if n not in unique_set or not sdfg.arrays[n].transient}
+        write_set = {n for n in write_set if n not in unique_set or not sdfg.arrays[n].transient}
+
+        # Find defined subgraph symbols
+        defined_symbols = set()
+        strictly_defined_symbols = set()
+        for e in subgraph.edges():
+            defined_symbols.update(set(e.data.assignments.keys()))
+            for k, v in e.data.assignments.items():
+                try:
+                    if k not in sdfg.symbols and k not in {str(a) for a in symbolic.pystr_to_symbolic(v).args}:
+                        strictly_defined_symbols.add(k)
+                except AttributeError:
+                    # `symbolic.pystr_to_symbolic` may return bool, which doesn't have attribute `args`
+                    pass
+
+        new_state = sdfg.add_state('nested_sdfg_parent')
+        nsdfg = SDFG("nested_sdfg", constants=sdfg.constants_prop, parent=new_state)
+        nsdfg.add_node(source_node, is_start_state=True)
+        nsdfg.add_nodes_from([s for s in states if s is not source_node])
+        for s in states:
+            s.parent = nsdfg
+        for e in subgraph.edges():
+            nsdfg.add_edge(e.src, e.dst, e.data)
+
+        for e in sdfg.in_edges(source_node):
+            sdfg.add_edge(e.src, new_state, e.data)
+        for e in sdfg.out_edges(sink_node):
+            sdfg.add_edge(new_state, e.dst, e.data)
+
+        sdfg.remove_nodes_from(states)
+
+        # Add NestedSDFG arrays
+        for name in read_set | write_set:
+            nsdfg.arrays[name] = copy.deepcopy(sdfg.arrays[name])
+            nsdfg.arrays[name].transient = False
+        for name in unique_set:
+            nsdfg.arrays[name] = sdfg.arrays[name]
+            del sdfg.arrays[name]
+
+        # Handle symbols taking new value inside NestedSDFG
+        ndefined_symbols = set()
+        out_mapping = {}
+        out_state = None
+        for e in nsdfg.edges():
+            ndefined_symbols.update(set(e.data.assignments.keys()))
+        if ndefined_symbols:
+            out_state = nsdfg.add_state('symbolic_output')
+            nsdfg.add_edge(sink_node, out_state, InterstateEdge())
+            for s in ndefined_symbols:
+                if s in nsdfg.symbols:
+                    dtype = nsdfg.symbols[s]
+                else:
+                    dtype = sdfg.symbols[s]
+                name, _ = sdfg.add_scalar(f"__sym_out_{s}", dtype, transient=True, find_new_name=True)
+                out_mapping[s] = name
+                nname, ndesc = nsdfg.add_scalar(f"__sym_out_{s}", dtype, find_new_name=True)
+                tasklet = out_state.add_tasklet(f"set_{nname}", {}, {'__out'}, f'__out = {s}')
+                acc = out_state.add_access(nname)
+                out_state.add_edge(tasklet, '__out', acc, None, Memlet.from_array(nname, ndesc))
+                write_set.add(name)
```

Comment:
what happens on the outer SDFG? how does this become a symbol outside later on?

Reply:
The symbol already exists in both the outer and the nested SDFG; it is part of the subgraph in the first place. We only need to add a scalar connector to pass the value from the nested SDFG to the outer one.